### ISSUE TITLE
OOMKill of bayesianoptimization pod

### ISSUE THREAD

**Issue by bataliero** [kubeflow/katib#2716]
### What happened

I ran a Katib Experiment using the `bayesianoptimization` algorithm (skopt) against
Trainer V2 TrainJob trials — 2 hyperparameters (learning rate, weight decay). After 17 trials completed
successfully, the `<experiment-name>-bayesianoptimization` suggestion Deployment
started CrashLoopBackOff. Every restart it ran for about 14 seconds then got
OOMKilled. The Experiment itself just sat stuck at `trialsSucceeded: 17`, no new
trials ever got proposed, no error visible anywhere except the suggestion pod's
own restart count climbing.

Checked the pod's resource limits and `katib-config`'s `bayesianoptimization`
entry has no `resources` block at all, so it falls back to whatever the
controller's built-in default is (looked like ~100Mi from testing). That's
apparently just fine only for the first dozen trials

### What fixed it

I added this to `katib-config`'s `bayesianoptimization` entry:

```yaml
- algorithmName: bayesianoptimization
  image: ghcr.io/kubeflow/katib/suggestion-skopt:v0.19.0
  resources:
    limits:
      memory: 512Mi
```
I propose updating default `https://github.com/kubeflow/katib/blob/master/manifests/v1beta1/installs/katib-cert-manager/katib-config.yaml` to 512Mi.

### Environment

Kubernetes version:
```bash
$ kubectl version
Client Version: v1.32.0
Kustomize Version: v5.5.0
Server Version: v1.35.6+rke2r1
WARNING: version difference between client (1.32) and server (1.35) exceeds the supported minor version skew of +/-1

```
Katib controller version:
```bash
$ kubectl get pods -n kubeflow -l katib.kubeflow.org/component=controller -o jsonpath="{.items[*].spec.containers[*].image}"
ghcr.io/kubeflow/katib/katib-controller:v0.19.0

```
Katib Python SDK version:
```bash
$ pip show kubeflow-katib

pip show kubeflow-katib
Name: kubeflow-katib
Version: 0.19.0
Summary: Katib Python SDK for APIVersion v1beta1
Home-page: https://github.com/kubeflow/katib/tree/master/sdk/python/v1beta1
Author: Kubeflow Authors
Author-email: premnath.vel@gmail.com
License: Apache License Version 2.0
Location: /home/pk/Documents/csem-ai-platform/.venv/lib/python3.12/site-packages
Requires: certifi, grpcio, kubeflow-training, kubernetes, protobuf, setuptools, six, urllib3
```

### Impacted by this bug?

Give it a 👍 We prioritize the issues with most 👍

**Comment by bataliero** [comment]
`Update 1`:
"tpe" and "multivariate-tpe" stayed essentially stable throughout the 100-trial run, with memory usage changing by only a few MB. In contrast, bayesianoptimization nearly tripled its memory usage in just 26 trials (78 MB → 220 MB).

`Update 2`:

Ran all 8 algorithms through the same 100-trial mock sweep:

| algorithm | memory after 100 trials |
|---|---|
| bayesianoptimization | 220 MB (only got to trial 32 - too slow/heavy to finish) |
| multivariate-tpe | 100 MB |
| grid | 54 MB |
| tpe | 92 MB |
| random | 91 MB |
| sobol | 13 MB |
| cmaes | 13 MB |
| hyperband | 79 MB (note: algo decided on its own that 8 trials was the whole experiment,) |


Every algorithm stayed quite flat except `bayesianoptimization`, which nearly tripled in the first 30 trials and crosses the 100Mi default around trial 10-14.

`multivariate-tpe` is on the edge of default 100Mi

**Comment by 100-JM** [comment]
Thanks for the detailed report. I opened #2720 for this.

Two findings while reproducing it against the `suggestion-skopt:v0.19.0` image driven like katib-controller (all Trials resent each call, 3 requests per call):

- I could not reproduce a growth past the first Trials: RSS stays flat at ~125-145MB from Trial 1 to 60 on Linux, with 1/10/64 OpenBLAS threads and under amd64 too. But the image already needs ~125MB right after start, so the 100Mi default is too small before any Trial runs. The PR sets `512Mi` in katib-config for `bayesianoptimization`, as you proposed.
- The "too slow/heavy" part is real and separate: OpenBLAS spins one thread per host core, and under the 500m CPU limit they get throttled together with the actual work. Pinning `OPENBLAS_NUM_THREADS=1` in the image makes 24 Trials at 0.5 CPU go from 97s to 5s.

If you still have the cluster, `kubectl top pod` on the suggestion pod right after it starts (before Trial 1) would be useful to confirm the baseline on your platform.

**Comment by bataliero** [comment]
Thanks for looking at this.

I re-ran it on our cluster with the stock `suggestion-skopt:v0.19.0` image, the
default 500m CPU limit that katib applies, `parallelTrialCount: 8`, and sampled
from inside the pod rather than via `kubectl top`.

**The baseline right after start, before the first trial:**

- stock image: RSS 129 MB, 83 threads
- with `OPENBLAS_NUM_THREADS=1`: RSS 128 MB, 21 threads

The suggestion calls get slower as trials accumulate: 2.9s, 6.7s, 17.4s, 25.3s.
Once a call crosses 60s it hits the controller's gRPC deadline
(`suggestionclient.go`), the controller gives up and retries — but the Python
servicer never checks whether the caller is still there, so the abandoned call
keeps computing. The log shows new calls starting at exactly 60s intervals with
no returns in between, and `main.py` uses `ThreadPoolExecutor(max_workers=10)`,
so they stack up.

Memory then tracks the number of stuck calls, not the trial count:

- 3 calls in flight: anon 215 MB
- 4 calls in flight: anon 233 MB, RSS 289 MB

roughly 15-18 MB per abandoned call, each holding its own `Optimizer.copy()`
plus BLAS scratch. At trial 32 the pod sat at RSS 272 MB / cgroup 224 MB, of
which anon was 215 MB and page cache 0.3 MB 

The container sees all 32 cores of the
node while being capped at half a core, so OpenBLAS sizes itself for 32 cores
and then spends its quota on contention: 1179 of 1351 periods throttled (87%),
about 6000s of cumulative throttled time, ~23 threads runnable at any instant.

Same experiment, same
node, only `OPENBLAS_NUM_THREADS=1` / `OMP_NUM_THREADS=1` added:

| | stock | threads=1 |
|---|---|---|
| threads | 83 | 21 |
| a `request_number=8` call | 25s, then never returns | 2.6s |
| CPU periods throttled | 87% | 13% |
| memory at 16 trials | 178 MB cgroup | 83 MB cgroup (RSS 137 MB) |
| outcome | permanently stalled at 32 trials | keeps going |

So on a many-core node the thread pinning is what actually prevents the failure;

Thanks!

**Comment by 100-JM** [comment]
Thanks, that explains it. The abandoned calls piling up after the 60s deadline is why memory tracked in-flight calls, not trials, and why I couldn't see it on a 10-core box where calls never crossed the deadline.

I added a commit to #2720 for that: `GetSuggestions` is now serialized with a lock (same as the optuna servicer) and aborts with `CANCELLED` if the caller already gave up, so a retry no longer computes on behalf of a caller that is gone. Thread pinning remains the main fix. Your numbers are quoted in the PR description. If you can try the branch image on that node, that would be the best confirmation.

**Comment by bataliero** [comment]
Hi 
Confirmed on our side!

I run it twice on same node
as before: 32 cores, default 500m suggestion CPU limit, `parallelTrialCount: 8`.


| | stock v0.19.0 | your branch |
|---|---|---|
| 60 trials | 38m16s | 3m11s / 3m48s |
| peak memory | 308 MB cgroup / 343 MB RSS | 105 and 124 MB / 172 and 178 MB |
| threads | 92 | 22 |
| slowest GetSuggestions | never returned, retried every 60s | 7.6s |


Worth noting for the peaks both over the **100Mi default.**  So with default katib configmap configuration OOMKill can be triggered.

Thanks!

**Comment by 100-JM** [comment]
Great, thank you for running it. I added your table to the PR description.

On the peaks still being above 100Mi: agreed, that is why the first commit in #2720 sets an explicit `512Mi` limit for `bayesianoptimization` in every katib-config, so the default install is covered too.
